filesystem/block.go

```diff
@@ -0,0 +1,340 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package main
+
+import (
+	"fmt"
+	"strings"
+)
+
+// BlockSize is the size of each data block. It must not
+// change throughout the lifetime of the filesystem.
+const BlockSize = 4 << 10 // 4KB
+
+func min(a, b uint64) uint64 {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// blockRange describes a range of blocks.
+// If the first and last block are the same, the effective data range
+// will be: [startOffset, lastLength)
+type blockRange struct {
+	start       int    // index of the start block
+	startOffset uint64 // starting offset within the first block
+	startLength uint64 // length of data in first block
+	last        int    // index of the last block
+	lastLength  uint64 // length of the last block
+}
+
+// newBlockRange returns the block range for 'size' bytes from 'from'.
+func newBlockRange(from, length uint64) blockRange {
+	end := from + length
+	return blockRange{
+		start:       int(from / BlockSize),
+		startOffset: from % BlockSize,
+		startLength: min(length, BlockSize-(from%BlockSize)),
+		last:        int(end / BlockSize),
+		lastLength:  end % BlockSize,
+	}
+}
+
+// shrink resizes the data to a smaller length.
+// Requirement: from > to.
+// If truncates are done on block boundaries, this is reasonably
+// efficient. However, if truncating in the middle of a block,
+// we need to fetch the block first, truncate it, and write it again.
+func shrink(e sqlExecutor, inodeID, from, to uint64) error {
+	delRange := newBlockRange(to, from-to)
+	deleteFrom := delRange.start
+
+	if delRange.startOffset > 0 {
+		// We're truncating in the middle of a block, fetch it, truncate its
+		// data, and write it again.
+		// TODO(marc): this would be more efficient if we had LEFT for bytes.
```

Comment:
I'll improve these in my next batch of work on this (ie: probably next week-end). My plan is to keep looking for neat stuff we could support, then someone implements them, then I use them.